### DISCUSSION TITLE
PyAT Developments

### DISCUSSION THREAD

**Discussion by lcarver** [atcollab/at#237] [Python AT]
Dear all, (@lfarv , @willrogers , @swhite2401 , @T-Nicholls , @carmignani )

I wanted to start a list of small improvements / bug fixes/ minor additional features to pyat. It seems github does not have the todo list capability (unlike gitlab) so a discussion is the most relevant place for it.

1. A function to determine the available lattice names inside a mat file.
2. Update the help to include twiss_in functionality. 
3. A function which takes twiss parameters as arguments and creates a recarray that can be used as input for twiss_in. 
4. A convenient set_cavity function. 
5. I started a PyAT tutorial but it is currently not finished and needs to be improved. I think we need to discuss about how we will encourage beginners to use PyAT. Is it through a well configured PyCharm? Or a more spartan version like I do where I use gedit and ipython? Once this is decided I can rewrite the tutorial to reflect this which can then be included in the repo.

I am sure there are others that I have forgotten. But once I remember them I can update them here. I am interested to know if you have any thoughts or comments on this.

Cheers,
Lee

**Comment by willrogers** [comment]
Sounds good to me. Consider writing tests as you go.

For the tutorial I don't think you should require people to download and set up Pycharm. I think Jupyter notebooks might be a good way of doing this - see https://github.com/atcollab/at/tree/master/jupyter.

**Comment by TeresiaOlsson** [comment]
I have some things I have noticed related to encouraging beginners to use pyAT in case it helps.

First, the need for a lattice file. To be able to import mat files is nice, but most people seem so used to having a text based lattice file that they can make changes to directly that they prefer to work that way. They don't want to first have to write a m file, produce a mat file and then import it. If they need to do that they just stick with using AT. They want a text based lattice file that can be read directly into pyAT. So in case that doesn't already exist now, I think a good step would be to implement that.

I haven't used pyCharm, but I use Spyder. Not sure how different the installation and set up is between the two, but I usually find Spyder to be easy. Jupyter notebooks are good for showing examples, but for a tutorial I think people would like examples of real simulations (many particles, many turns etc) that they can translate to their lattice and I don't think anyone does that using Jupyter? One advantage with AT/pyAT over other codes is the possibility to look at variables during the execution, which you can do in Spyder in the same way as in Matlab and I think it would be good to highlight that.

Otherwise I think Elegant is a good example. Elegant has a very steep learning curve, but they still attract a lot of new users. Partly because of all the functionality they have implemented, but I think also because of the resources available for the users. Things like a manual, user forum and a lot of examples that you can download and modify for your machine. Those things take a lot of time to build up, but perhaps with a good structure for how to do it pyAT can also get there with time? Especially having a user forum I think could be useful because then more people can contribute to provide support for new users.

**Reply by willrogers** [reply]
You can load Tracy and Elegant files into pyat. I doubt that the functionality is complete but I can extend it if necessary.

**Reply by willrogers** [reply]
Also, I'm not sure that you can't do the examples that you describe using Jupyter. We could try.

**Reply by TeresiaOlsson** [reply]
> You can load Tracy and Elegant files into pyat. I doubt that the functionality is complete but I can extend it if necessary.

That is very good, but in in addition to that I think people really want to be able to write a lattice file directly for pyAT. When I tell them that there isn't a lattice file, but they can read in files from a lot of other codes they get all confused and don't understand why there isn't a pyAT lattice file. I think this might be an area where people are just so used to how things should work that they want it to work that way. Also, in case you want to use pyAT because it doesn't require a license and is easy to install, you might not have access to other codes where you can build the lattice file since that might require to run it in those codes first to check that the syntax is correct. I think having a pyAT lattice file would make it more of a standalone code.

**Reply by willrogers** [reply]
I think it would make no sense at all to invent a new lattice file format. We could adopt one of the other formats 'officially' if that would help. Are any of the formats well documented?

**Reply by TeresiaOlsson** [reply]
I think using the same format as AT, but just be able to write it in a py file instead of m file and read it in directly without having to first create a mat file would be the best solution. So the AT format, but for Python instead of Matlab?

**Reply by lcarver** [reply]
Hello Teresia,

The lattice files are a tricky topic but I think that we should maintain some compatibility with AT. Nonetheless for me I would like to try and completely remove the matlab aspects that are still present in PyAT and to me this seems to be the largest matlab component still lingering.  

However to me it is more that we have not spent the time to go through the fundamental lattice creation for our lattices, which would then be saved as a mat file (in fact there is the option for saving as an m-file, maybe an additional extension could be created for python that has more readability). Right now it is just more convenient to use the mat files that already exist. 

In fact in the tutorial that I started developing, I create a simple DBA lattice from scratch so it is not a difficult exercise but for a full realistic machine it is a time consuming exercise.

**Reply by matsievskiysv** [reply]
[YAML](https://pyyaml.org/) is good for textual serialization of Python objects. There's also a Matlab/Octave [version](https://github.com/biogeo/mat-yaml).

**Reply by TeresiaOlsson** [reply]
Mm. I know. Creating lattice files are usually a time consuming mess, but being able to create a lattice from scratch I think is necessary to get new users. Otherwise only people that already have a lattice will be able to use it and then they might not see the point of spending time to switch to another code. That is one of the reasons why we haven't started to use pyAT in the AP group at Diamond. I think pyAT also has a lot of potential to be used by students (which increasingly seem to have less access to a Matlab license through their university) and they I think need the option to build a lattice from scratch.

How did you build the lattice from scratch in the tutorial? I was hoping that it should be possible to just translate the way AT does it into python syntax, but perhaps it's more complicated than that?

**Reply by willrogers** [reply]
You can create lattice files directly in Python already, but text-based formats are much better for long-term management.

There's nothing special about loading Matlab formats in pyat, it is just one of the formats that we can load.

**Reply by TeresiaOlsson** [reply]
@lcarver sent me the tutorial. It was great. I think that's exactly what people would want. The tutorial explains how to create a lattice using the elements definitions, combine it into a lattice and save as a mat file that can be imported when doing a simulation without the need for having other codes. That is what I meant by a pyAT lattice file.

**Comment by lfarv** [comment]
`pyat` has a very simple and native text-based lattice representation !

It's similar to the Matlab .m lattice file, it's just a list of python commands as they would be interpreted at the command line.
You just list the element constructors, one per line.
You can use it with:
```
>> ring = at.load_lattice('test.repr')
>> at.save_lattice(ring, 'test.repr')
```
Here is an example corresponding to the 'hmba.mat' example (remove the .txt extension): [test.repr.txt]
(https://github.com/atcollab/at/files/6149917/test.repr.txt)
I created it with the sequence:
```
>> import at
>> ring=at.load_lattice('<atroot>/pyat/test_matlab/hmba.mat')
>> ring.radiation_off(quadrupole_pass='auto')
>> at.save_lattice(ring,'test.repr')
```
Note that is has a lot of unnecessary attributes inherited from the .mat file, you just need to specify the non-default attributes.

However the .repr format is very simple and does not offer the possibilities of other languages.

**Reply by TeresiaOlsson** [reply]
I think that is what the people I have talked to is asking for. They want to be able to build a lattice in that way and not only have the option to import already existing files. I didn't know that this option existed.

**Reply by lfarv** [reply]
Few people know… That's why I am happy that someone starts building a tutorial!
Go on @lcarver!
